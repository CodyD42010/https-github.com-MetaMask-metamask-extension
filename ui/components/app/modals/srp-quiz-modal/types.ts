import { BUTTON_VARIANT } from 'ui/components/component-library';

export enum QuizStage {
  introduction = 'introduction',
  questionOne = 'question-one',
  wrongAnswerQuestionOne = 'wrong-answer-question-one',
  rightAnswerQuestionOne = 'right-answer-question-one',
  questionTwo = 'question-two',
  wrongAnswerQuestionTwo = 'wrong-answer-question-two',
  rightAnswerQuestionTwo = 'right-answer-question-two',
}

export interface IQuizInformationProps {
  title: {
    content: string;
    style?: any;
  };
  buttons: {
    onClick: () => void;
    label: string;
    variant: string;
  }[];
  dismiss: () => void;
  header?: string;
  image?: string; // was ImageSourcePropType
  content?: string;
  icon?: any;
}
