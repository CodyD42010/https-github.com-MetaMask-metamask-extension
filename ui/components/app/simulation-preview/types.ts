import { Hex } from '@metamask/utils';
import { TokenStandard } from '../../../../shared/constants/transaction';

/**
 * Identifies the native asset of a chain.
 */
export type NativeAssetIdentifier = {
  standard: TokenStandard.none;
  address?: undefined;
  tokenId?: undefined;
};

/**
 * Uniquely identifies an ERC20 token on a chain.
 */
export type Erc20AssetIdentifier = {
  standard: TokenStandard.ERC20;
  address: Hex;
  tokenId?: undefined;
};

/**
 * Uniquely identifies an ERC721 token on a chain.
 */
export type Erc721AssetIdentifier = {
  standard: TokenStandard.ERC721;
  address: Hex;
  tokenId: Hex;
};

/**
 * Uniquely identifies an ERC1155 token on a chain.
 */
export type Erc1155AssetIdentifier = {
  standard: TokenStandard.ERC1155;
  address: Hex;
  tokenId: Hex;
};

/**
 * Uniquely identifies an asset on a chain.
 */
export type AssetIdentifier =
  | NativeAssetIdentifier
  | Erc20AssetIdentifier
  | Erc721AssetIdentifier
  | Erc1155AssetIdentifier;

/**
 * Represents an amount of an asset, including its magnitude and sign.
 */
export type Amount = {
  /**
   * Indicates whether the amount is negative (e.g., a decrease in balance).
   */
  isNegative: boolean;

  /**
   * The quantity of the smallest denomination of the asset (base units),
   * represented as a hexadecimal string.
   * For example: In the case of ETH, this would be the number of wei.
   */
  quantity: Hex;

  /**
   * The number of decimal places the associated asset supports.
   *
   * This value is the negation of the exponent used when converting
   * the quantity to the decimal amount of a token.
   *
   * To calculate the token amount in decimal form, use the formula:
   * `tokenAmount = hexToDecimal(quantity) / (10 ^ decimals)`
   *
   * Example: If the asset is ETH, the quantity is expressed in wei
   * (the smallest unit of ETH) and decimals would be 18. The amount
   * of ETH tokens would be: `ethAmount = quantity / (10 ^ 18)`
   */
  decimals: number;
};

/**
 * Describes a change in an asset's balance to a user's wallet.
 */
export type BalanceChange = {
  asset: AssetIdentifier;
  amount: Amount;
};
