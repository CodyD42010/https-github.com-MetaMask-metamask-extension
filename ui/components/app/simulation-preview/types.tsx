import { Hex } from '@metamask/utils';
import { Numeric } from '../../../../shared/modules/Numeric';

/**
 * Specifies an asset.
 */
export type AssetInfo = {
  isNative: boolean;
  contractAddress?: Hex;
  tokenId?: Hex;
};

/**
 * A balance change for an asset.
 */
export type BalanceChange = {
  assetInfo: AssetInfo;
  isDecrease: boolean;
  absChange: Numeric;
};
