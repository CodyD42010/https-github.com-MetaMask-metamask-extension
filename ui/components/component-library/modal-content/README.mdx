import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import { ModalContent } from './modal-content';

# ModalContent

`ModalContent` is the container for the modal dialog's content. It uses context supplied by the `Modal` component and cannot be used without it.

[MetaMask Design System Guides](https://www.notion.so/MetaMask-Design-System-Guides-Design-f86ecc914d6b4eb6873a122b83c12940)

<Canvas>
  <Story id="components-componentlibrary-modalcontent--default-story" />
</Canvas>

## Props

<ArgsTable of={ModalContent} />

### Children

Use the `children` prop to render the content of `ModalContent`. The `ModalContent` should generally be used with the `ModalHeader` component.

<Canvas>
  <Story id="components-componentlibrary-modalcontent--children" />
</Canvas>

```jsx
import React, { useState } from 'react';
import { Modal, ModalContent, ModalHeader, Text, Button, ButtonVariant } from '../../component-library';

const [show, setShow] = useState(false);
const handleOnClick = () => {
  setShow(!show);
};

<Button variant={ButtonVariant.Primary} onClick={handleOnClick}>
  Open
</Button>
<Modal isOpen={show} onClose={handleOnClick}>
  <ModalContent {...args}>
    <ModalHeader marginBottom={4}>Modal Header</ModalHeader>
    <Text marginBottom={4}>Modal Content</Text>
    <Button variant={ButtonVariant.Primary} onClick={handleOnClick}>
      Close
    </Button>
    <LoremIpsum />
    <LoremIpsum />
    <LoremIpsum />
    <LoremIpsum />
    <LoremIpsum />
  </ModalContent>
</Modal>
```

### Size

Currently the `ModalContent` supports a single size, this decision was made after we ran an audit on all modal sizes in the extension codebase and found that all modals could be made to fit the `ModalContentSize.Sm`(360px) size.

If you do require a larger modal size you can use the `modalDialogProps` to add a className to override the default size.

<Canvas>
  <Story id="components-componentlibrary-modalcontent--size" />
</Canvas>

```jsx
import { Display } from '../../../helpers/constants/design-system';

import { Box, Modal, ModalContent, Text, Button, ButtonVariant } from '../../component-library';

enum ModalContentSizeStoryOption {
  Sm = 'sm',
  ClassName = 'className',
}

const [show, setShow] = useState({
  sm: false,
  className: false,
});
const handleOnClick = (size: ModalContentSizeStoryOption) => {
  setShow({ ...show, [size]: !show[size] });
};

<Box display={Display.Flex} gap={4}>
  <Button
    variant={ButtonVariant.Secondary}
    onClick={() => handleOnClick(ModalContentSizeStoryOption.Sm)}
  >
    Show sm size
  </Button>
  <Button
    variant={ButtonVariant.Secondary}
    onClick={() => handleOnClick(ModalContentSizeStoryOption.ClassName)}
  >
    Show className
  </Button>
</Box>

<Modal
  isOpen={show.sm}
  onClose={() => handleOnClick(ModalContentSizeStoryOption.Sm)}
>
  <ModalContent {...args}>
    <Text marginBottom={4}>
      ModalContentSize.Sm default and only size 360px max-width
    </Text>
    <Button onClick={() => setShow({ ...show, sm: false })}>Close</Button>
  </ModalContent>
</Modal>

<Modal
  isOpen={show.className}
  onClose={() => handleOnClick(ModalContentSizeStoryOption.ClassName)}
>
  <ModalContent
    {...args}
    modalDialogProps={{
      style: { maxWidth: 800 },
    }}
  >
    <Text marginBottom={4}>
      Using modalDialogProps and adding a className setting a max width
      (max-width: 800px)
    </Text>
    <Button onClick={() => setShow({ ...show, className: false })}>
      Close
    </Button>
  </ModalContent>
</Modal>
```
