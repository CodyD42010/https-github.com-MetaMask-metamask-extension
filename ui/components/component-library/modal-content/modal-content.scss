.mm-modal-content {
  position: fixed;
  left: 0;
  top: 0;
  z-index: $modal-z-index;
  overflow: auto;
  overscroll-behavior-y: none;

  // Scroll bar styles
  // Firefox
  scrollbar-width: thin;
  scrollbar-color: var(--color-icon-muted) transparent;

  // Webkit: Chrome, Brave
  ::-webkit-scrollbar {
    width: 8px;
  }

  ::-webkit-scrollbar-thumb {
    -webkit-border-radius: 8px;
    border-radius: 8px;
    background: var(--color-icon-muted);
  }

  &__dialog {
    --modal-content-size: var(--size, 360px);

    max-height: 100%;
    box-shadow: var(--shadow-size-lg) var(--color-shadow-default);
    overflow-y: auto;  // TODO: Remove this once all Modals are using ModalBody correctly ideally this scroll bar should not be needed

    &--size-sm {
      --size: 360px;

      max-width: var(--modal-content-size);
    }

    &--size-md {
      --size: 480px;

      max-width: var(--modal-content-size);
    }

    &--size-lg {
      --size: 720px;

      max-width: var(--modal-content-size);
    }
  }
}
