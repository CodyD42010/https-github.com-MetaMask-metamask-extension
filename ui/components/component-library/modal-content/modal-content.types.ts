import React from 'react';
import type { BoxProps } from '../../ui/box/box.d';

/*
 * ModalContent sizes
 * Currently there is only use case for one size of ModalContent in the extension
 * See audit https://www.figma.com/file/hxYqloYgmVcgsoiVqmGZ8K/Modal-Audit?node-id=481%3A244&t=XITeuRB1pRc09hiG-1
 * Not to say there won't be more in the future, but to prevent redundant code there is only one for now
 */
export enum ModalContentSize {
  Sm = 'sm',
}

export interface ModalContentProps extends BoxProps {
  /**
   * The additional className of the ModalContent component
   */
  className?: string;
  /**
   * The content of the ModalContent component
   */
  children: React.ReactNode;
  /**
   * The size of ModalContent
   * Currently only one size is supported ModalContentSize.Sm 360px
   * See docs for more info
   */
  size?: ModalContentSize;
  /**
   * Additional props to pass to the dialog node inside of ModalContent component
   */
  modalDialogProps?: BoxProps | React.HTMLAttributes<HTMLElement>;
}
