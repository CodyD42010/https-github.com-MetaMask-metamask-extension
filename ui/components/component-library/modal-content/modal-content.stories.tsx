import React, { useState } from 'react';
import { StoryFn, Meta } from '@storybook/react';

import { Display, FlexWrap } from '../../../helpers/constants/design-system';

import {
  Box,
  ButtonVariant,
  Button,
  Text,
  Modal,
  ModalHeader,
  ModalBody,
} from '..';

import { ModalContent } from './modal-content';
import { ModalContentSize } from './modal-content.types';

import README from './README.mdx';

export default {
  title: 'Components/ComponentLibrary/ModalContent',
  component: ModalContent,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    children: { control: 'text' },
    size: {
      control: 'select',
      options: Object.values(ModalContentSize),
    },
    modalDialogProps: { control: 'object' },
  },
} as Meta<typeof ModalContent>;

const LoremIpsum = () => (
  <Text marginBottom={4}>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam euismod
    tortor vitae nisi blandit, eu aliquam nisl ultricies. Donec euismod
    scelerisque nisl, sit amet aliquet nunc. Donec euismod, nisl vitae
    consectetur aliquam, nunc nunc ultricies nunc, eget aliquam nisl nisl vitae
    nunc. Donec euismod, nisl vitae consectetur aliquam, nunc nunc ultricies
    nunc, eget aliquam nisl nisl vitae nunc. Donec euismod, nisl vitae
    consectetur aliquam, nunc nunc ultricies nunc, eget aliquam nisl nisl vitae
    nunc. Donec euismod, nisl vitae consectetur aliquam, nunc
  </Text>
);

export const DefaultStory: StoryFn<typeof ModalContent> = (args) => {
  const [show, setShow] = useState(false);
  const handleOnClick = () => {
    setShow(!show);
  };
  return (
    <>
      <Button variant={ButtonVariant.Primary} onClick={handleOnClick}>
        Open
      </Button>
      <Modal isOpen={show} onClose={handleOnClick}>
        <ModalContent {...args}>
          <ModalHeader>Modal Header</ModalHeader>
          <ModalBody>
            <Text>Modal Content</Text>
          </ModalBody>
          <Box padding={4}>
            <Button
              variant={ButtonVariant.Primary}
              onClick={handleOnClick}
              block
            >
              Close
            </Button>
          </Box>
        </ModalContent>
      </Modal>
    </>
  );
};

DefaultStory.storyName = 'Default';

export const Children: StoryFn<typeof ModalContent> = (args) => {
  const [show, setShow] = useState(false);
  const handleOnClick = () => {
    setShow(!show);
  };
  return (
    <>
      <Button variant={ButtonVariant.Primary} onClick={handleOnClick}>
        Open
      </Button>
      <Modal isOpen={show} onClose={handleOnClick}>
        <ModalContent {...args}>
          <ModalHeader marginBottom={4}>Modal Header</ModalHeader>
          <ModalBody>
            <Text marginBottom={4}>
              The ModalContent with ModalHeader and Text components as children
            </Text>
            <Button
              marginBottom={4}
              variant={ButtonVariant.Primary}
              onClick={handleOnClick}
            >
              Close
            </Button>
            <LoremIpsum />
            <LoremIpsum />
            <LoremIpsum />
            <LoremIpsum />
            <LoremIpsum />
          </ModalBody>
        </ModalContent>
      </Modal>
    </>
  );
};

enum ModalContentSizeStoryOption {
  Sm = 'sm',
  Md = 'md',
  Lg = 'lg',
  ClassName = 'className',
}

export const Size: StoryFn<typeof ModalContent> = (args) => {
  const [currentSize, setCurrentSize] =
    useState<ModalContentSizeStoryOption | null>(null);

  const handleOnClick = (size: ModalContentSizeStoryOption) => {
    setCurrentSize(currentSize === size ? null : size);
  };

  const getSize = (size: ModalContentSizeStoryOption) => {
    switch (size) {
      case ModalContentSizeStoryOption.Sm:
        return ModalContentSize.Sm;
      case ModalContentSizeStoryOption.Md:
        return ModalContentSize.Md;
      case ModalContentSizeStoryOption.Lg:
        return ModalContentSize.Lg;
      default:
        return ModalContentSize.Sm;
    }
  };

  return (
    <>
      <Box display={Display.Flex} flexWrap={FlexWrap.Wrap} gap={4}>
        {Object.values(ModalContentSizeStoryOption).map((size) => (
          <Button
            key={size}
            variant={ButtonVariant.Secondary}
            onClick={() => handleOnClick(size)}
          >
            {`Show ${size} size`}
          </Button>
        ))}
      </Box>
      {currentSize && (
        <Modal isOpen={true} onClose={() => setCurrentSize(null)}>
          <ModalContent
            {...args}
            size={getSize(currentSize)}
            modalDialogProps={
              currentSize === ModalContentSizeStoryOption.ClassName
                ? { className: 'max-width-800' }
                : {}
            }
          >
            <ModalHeader marginBottom={4} onClose={() => setCurrentSize(null)}>
              {`ModalContent size: ${currentSize}`}
            </ModalHeader>
            <ModalBody>
              <Text marginBottom={4}>
                {currentSize === ModalContentSizeStoryOption.ClassName ? (
                  <>
                    This ModalContent has size set using modalDialogProps and
                    adding a className setting a max width (max-width: 800px){' '}
                  </>
                ) : (
                  <>This ModalContent is using size: {currentSize}</>
                )}
              </Text>
            </ModalBody>
            <Box padding={4}>
              <Button onClick={() => setCurrentSize(null)} block>
                Close
              </Button>
            </Box>
          </ModalContent>
        </Modal>
      )}
    </>
  );
};
