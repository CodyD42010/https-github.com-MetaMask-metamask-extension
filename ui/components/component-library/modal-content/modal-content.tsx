import React, { forwardRef, useRef, useEffect } from 'react';
import classnames from 'classnames';

import {
  BackgroundColor,
  BorderRadius,
  BLOCK_SIZES,
  DISPLAY,
  JustifyContent,
  AlignItems,
} from '../../../helpers/constants/design-system';

import Box from '../../ui/box/box';

import { ModalFocus, useModalContext } from '..';

import { ModalContentProps, ModalContentSize } from './modal-content.types';

export const ModalContent = forwardRef(
  (
    {
      className = '',
      children,
      size = ModalContentSize.Sm,
      modalDialogProps,
      ...props
    }: ModalContentProps,
    ref: React.Ref<HTMLElement>,
  ) => {
    const {
      onClose,
      isClosedOnEscapeKey,
      isClosedOnOutsideClick,
      initialFocusRef,
      finalFocusRef,
      restoreFocus,
      autoFocus,
    } = useModalContext();
    const modalDialogRef = useRef<HTMLElement>(null);
    const handleEscKey = (event: KeyboardEvent) => {
      if (isClosedOnEscapeKey && event.key === 'Escape') {
        onClose();
      }
    };

    const handleClickOutside = (event: MouseEvent) => {
      if (
        isClosedOnOutsideClick &&
        modalDialogRef?.current &&
        !modalDialogRef.current.contains(event.target as Node)
      ) {
        onClose();
      }
    };

    useEffect(() => {
      document.addEventListener('keydown', handleEscKey);
      document.addEventListener('mousedown', handleClickOutside);

      return () => {
        document.removeEventListener('keydown', handleEscKey);
        document.removeEventListener('mousedown', handleClickOutside);
      };
    }, []);
    return (
      <ModalFocus
        initialFocusRef={initialFocusRef}
        finalFocusRef={finalFocusRef}
        restoreFocus={restoreFocus}
        autoFocus={autoFocus}
      >
        <Box
          className={classnames('mm-modal-content', className)}
          ref={ref}
          display={DISPLAY.FLEX}
          width={BLOCK_SIZES.SCREEN}
          height={BLOCK_SIZES.SCREEN}
          justifyContent={JustifyContent.center}
          alignItems={AlignItems.flexStart}
          padding={4}
          {...props}
        >
          <Box
            className={classnames(
              'mm-modal-content__dialog',
              `mm-modal-content__dialog--size-${size}`,
            )}
            as="section"
            role="dialog"
            aria-modal="true"
            backgroundColor={BackgroundColor.backgroundDefault}
            borderRadius={BorderRadius.LG}
            width={BLOCK_SIZES.FULL}
            marginTop={12}
            marginBottom={12}
            padding={4}
            ref={modalDialogRef}
            {...modalDialogProps}
          >
            {children}
          </Box>
        </Box>
      </ModalFocus>
    );
  },
);
