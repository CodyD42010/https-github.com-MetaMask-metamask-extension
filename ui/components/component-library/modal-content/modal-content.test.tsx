/* eslint-disable jest/require-top-level-describe */
import { render } from '@testing-library/react';
import React from 'react';

import { ModalContent } from './modal-content';
import { ModalContentSize } from './modal-content.types';

describe('ModalContent', () => {
  it('should render with text inside the ModalContent', () => {
    const { getByText } = render(<ModalContent>test</ModalContent>);
    expect(getByText('test')).toBeDefined();
    expect(getByText('test')).toHaveClass('mm-modal-content');
  });
  it('should match snapshot', () => {
    const { container } = render(<ModalContent>test</ModalContent>);
    expect(container).toMatchSnapshot();
  });
  it('should render with and additional className', () => {
    const { getByText } = render(
      <ModalContent className="test-class">test</ModalContent>,
    );
    expect(getByText('test')).toHaveClass('test-class');
  });
  it('should render with size sm', () => {
    const { getByText } = render(
      <>
        <ModalContent>default</ModalContent>
        <ModalContent size={ModalContentSize.Sm}>sm</ModalContent>
      </>,
    );
    expect(getByText('sm')).toHaveClass('mm-modal-content--size-sm');
    expect(getByText('default')).toHaveClass('mm-modal-content--size-sm');
  });
  it('should render with a ref', () => {
    const ref = React.createRef<HTMLDivElement>();
    render(<ModalContent modalContentRef={ref}>test</ModalContent>);
    expect(ref.current).toBeDefined();
  });
});
