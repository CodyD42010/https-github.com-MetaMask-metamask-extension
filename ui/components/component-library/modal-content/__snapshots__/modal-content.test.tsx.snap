// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalContent should match snapshot 1`] = `
<div
  class="box mm-modal-content box--padding-4 box--display-flex box--flex-direction-row box--justify-content-center box--align-items-flex-start box--width-screen box--height-screen"
  data-testid="test"
>
  <section
    aria-modal="true"
    class="box mm-modal-content__dialog mm-modal-content__dialog--size-sm box--margin-top-12 box--margin-bottom-12 box--padding-4 box--flex-direction-row box--width-full box--background-color-background-default box--rounded-lg"
    role="dialog"
  >
    test
  </section>
</div>
`;
